modules/platforms/cpp/schema/BinaryTupleBuilder.h

```diff
@@ -0,0 +1,378 @@
+        if (slice.has_value()) {
+            append(type, slice.value());
+        } else {
+            append(std::nullopt);
+        }
+    }
+
+    /**
+     * @brief appends values for a number of elements
+     *
+     * @tparam BytesT Byte container for a single internal tuple field.
+     * @param schema Tuple schema.
+     * @param tuple Tuple in the internal form.
+     */
+    template <typename BytesT>
+    void append(const BinaryTupleSchema &schema, const std::vector<std::optional<BytesT>> &tuple) {
+        for (IntT i = 0; i < schema.numElements(); i++) {
+            append(schema.getElement(i).dataType, tuple[i]);
+        }
+    }
+
+    /**
+     * @brief finalizes and returns a binary tuple
+     *
+     * @return Byte buffer with binary tuple.
+     */
+    const std::vector<std::byte> &build() {
+        assert(elementIndex == elementCount);
+        return binaryTuple;
+    }
+
+    /**
+     * @brief builds a binary tuple from an internal tuple representation
+     *
+     * @tparam BytesT Byte container for a single internal tuple field.
+     * @param schema Tuple schema.
+     * @param tuple Tuple in the internal form.
+     * @return Byte buffer with binary tuple.
+     */
+    template <typename BytesT>
+    const std::vector<std::byte> &build(const BinaryTupleSchema &schema, const std::vector<std::optional<BytesT>> &tuple) {
+        start();
+        claim(schema, tuple);
+        layout();
+        append(schema, tuple);
+        return build();
+    }
+
+private:
+    /**
+     * @brief computes required binary size for a given value
+     *
+     * @param value Actual element value.
+     * @return Required size.
+     */
+    static SizeT sizeOfInt8(std::int8_t value) noexcept {
+        return value == 0 ? 0 : sizeof(std::int8_t);
+    }
+
+    /**
+     * @brief computes required binary size for a given value
+     *
+     * @param value Actual element value.
+     * @return Required size.
+     */
+    static SizeT sizeOfInt16(std::int16_t value) noexcept {
+        if (std::uint16_t(value) <= UINT8_MAX) {
+            return sizeOfInt8(std::int8_t(value));
+        }
+        return sizeof(std::int16_t);
+    }
+
+    /**
+     * @brief computes required binary size for a given value
+     *
+     * @param value Actual element value.
+     * @return Required size.
+     */
+    static SizeT sizeOfInt32(std::int32_t value) noexcept {
+        if (std::uint32_t(value) <= UINT8_MAX) {
+            return sizeOfInt8(std::int8_t(value));
+        }
+        if (std::uint32_t(value) <= UINT16_MAX) {
+            return sizeof(std::int16_t);
+        }
+        return sizeof(std::int32_t);
+    }
+
+    /**
+     * @brief computes required binary size for a given value
+     *
+     * @param value Actual element value.
+     * @return Required size.
+     */
+    static SizeT sizeOfInt64(std::int64_t value) noexcept {
+        if (std::uint64_t(value) <= UINT16_MAX) {
```

Comment:
How does this work? In Java code I see:
```        
if (Short.MIN_VALUE <= value && value <= Short.MAX_VALUE) {
    return appendShort((short) value);
}
```
But here I see a conversion to unsigned type. I really doubt that `std::uint64_t` returns an absolute value, or does it? If it does, then I'd rather see an explicit `abs` call. If it doesn't, then this code doesn't seem safe, or even correct at all.

Reply:
Fixed, thanks.